### DISCUSSION TITLE
Why does xyflow use arrays for nodes and edges instead of hashmaps?

### DISCUSSION THREAD

**Discussion by Randulfe** [xyflow/xyflow#3894] [General]
### Please describe the feature that you want to propose

On my flow I usually have to update, or retrieve data from specific nodes multiple times and quite frequently which requires me to very innefficiently iterate over the whole node or edge array to find the specific node I need to update or compare the data with. Is it possible to instead switch to hash maps to find nodes more efficiently? And if it is, how can compatible is this data structure really with the internals and API methods of react flow?

**Comment by Spriithy** [comment]
You can always use an object to store the `{ [id]: indexInArray }` pairs

**Comment by bcakmakoglu** [comment]
Are you aware that there's a function called [`getNode`](https://reactflow.dev/api-reference/types/react-flow-instance#get-node) that doesn't need you to iterate over all nodes to find a specific one?

```ts
const { getNode } = useReactFlow()

const doSomething = () => {
   const node = getNode(nodeId)

	if (node) {
		// do something...
	}
}

**Comment by moklick** [comment]
As @bcakmakoglu explained, `getNode` could help. It uses the internal nodes Map (`nodeInternals`). You can also access it directly by using the `useStore` hook:

```js
const node = useStore(s => s.nodeInternals.get(nodeId));
```
